### PR TITLE
JAMES-420 correct hasAttachment query

### PR DESCRIPTION
**Pull request by strokyl**
JAMES-420 hasAttachment query (this PR is not WIP anymore)

### CODE REVIEW

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java [hunk 1]
```diff
@@ -68,12 +68,12 @@ public static IndexableMessage from(MailboxMessage message, List<User> users, Te
...
-        if (indexAttachments.equals(IndexAttachments.YES)) {
-            attachments = parsingResult.getAttachmentsStream()
-                    .collect(Collectors.toList());
-        } else {
-            attachments = ImmutableList.of();
-        }
+        List<MimePart> mimeparts = parsingResult.getAttachmentsStream()
+                .collect(Collectors.toList());
+
+        hasAttachment = mimeparts.stream().anyMatch(IndexableMessage::isAttachmentMimepart);
```
Comment (strokyl) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: I don't know if I should separate this method in two method one for compute hasAttachment the other one for setting attachment

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java [hunk 2]
```diff
@@ -155,11 +162,12 @@ private void generateText() {
     private Optional<String> bodyHtml;
     private String text;
 
+
     public IndexableMessage(long uid, String mailboxId, List<String> users, long modSeq, long size, String date, String mediaType,
```
Comment (strokyl) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: Arf that is horrible, can I change that to a builder or something like that ?

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java [hunk 3]
```diff
@@ -523,6 +528,18 @@ private Date toISODate(String value) throws ParseException {
         return cal.getTime();
     }
 
+
+    private boolean matches(SearchQuery.AttachmentCriterion criterion, MailboxMessage message) throws UnsupportedSearchException {
+        boolean mailHasAttachments = message.getAttachments().stream().anyMatch(new Predicate<MessageAttachment>() {
```
Comment (strokyl) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java]: I will refactor that there is the same logic for lucene

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 1 above)
Comment (mbaechler) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: hasAttachment is a "view" of attachments so it's not a problem to set both here.
The question here : do you really need hasAttachment to be a field ? I think that a method would do.

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java [hunk 4]
```diff
@@ -68,12 +68,12 @@ public static IndexableMessage from(MailboxMessage message, List<User> users, Te
...
-                    .collect(Collectors.toList());
-        } else {
-            attachments = ImmutableList.of();
-        }
+        List<MimePart> mimeparts = parsingResult.getAttachmentsStream()
+                .collect(Collectors.toList());
+
+        hasAttachment = mimeparts.stream().anyMatch(IndexableMessage::isAttachmentMimepart);
+
+        this.attachments = IndexAttachments.YES.equals(indexAttachments) ? mimeparts : ImmutableList.of();
```
Comment (mbaechler) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: we don't use ternary operator

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 2 above)
Comment (mbaechler) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: that's a private thing, isn't it ?

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 1 above)
Comment (strokyl) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: @mbaechler it was that or either I have to store indexAttachments inside the IndexableMessage, and that what I did before but this lead to less logical code IMO.

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 4 above)
Comment (strokyl) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: Why ? That is sad I love them.

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 2 above)
Comment (strokyl) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: Not totally check the class IndexableMessageWithMessageId

Comment (chibenwa) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: That's maybe a lot of work, but it can be cool. Try to do it as a separate commit, and don't invest too much time on it ;-)

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java (hunk 3 above)
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java]: You can have a helper class in mailbox/store that does this.

However I think you should iterate on the all body InputStream using Mime4J (a bit like done in ES) as some implementation, like JPA, don't support attachment retrieval.

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 4 above)
Comment (chibenwa) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: They are not readable

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java [hunk 5]
```diff
@@ -80,28 +81,28 @@
 
             /**
              * addr-mailbox of the first "cc" address.
-             * 
+             *
```
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java]: Sorry, but you have lot's of noise here...

Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java]: If you want to remove trailing spaces, do that in a separate commit.

Please also unset your IDE to avoid side effects upon save / commit...

File: mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java [hunk 6]
```diff
@@ -430,23 +438,23 @@ protected Analyzer createAnalyzer(boolean lenient) {
...
+     *
      * @param suffixMatch
      */
     public void setEnableSuffixMatch(boolean suffixMatch) {
         this.suffixMatch = suffixMatch;
     }
-    
-    
-    
+
```
Comment (chibenwa) [inline mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java]: Really, reformat in commits without code changes, please....

It would make review 100x easier

File: mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java [hunk 7]
```diff
@@ -739,23 +748,32 @@ public void body(BodyDescriptor desc, InputStream in) throws MimeException, IOEx
...
             throw new MailboxException("Unable to index content of message", e);
         }
-       
+
 
         return doc;
     }
 
+    private static boolean hasAttachment(MailboxMessage membership) {
+        return membership.getAttachments().stream().anyMatch(new Predicate<MessageAttachment>() {
```
Comment (chibenwa) [inline mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java]: No stream here, it's a java 6 project for now...

You can use Gava's FluentIterable instead.

Also, extract your predicate in an other method to make reading more easy.

Comment (chibenwa) [inline mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java]: Skip more lines (btw stream and anyMatch)

And thanks I did not knew anyMatch ;-)

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java (hunk 3 above)
Comment (chibenwa) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java]: More line break PLZ :-)

And extract predicate

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 1 above)
Comment (chibenwa) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: I prefer having hasAttachment as a field : denormalization, you know... ES will search this more easily, and the actual ES query won't have to do a filter / find any for every messages ;-)

**Comment by chibenwa** [comment]
I propose you there 3 commits : 
 - https://github.com/chibenwa/james-project/commit/6051f7fc17aa16397cd642d55583bf35bf9e5b10 StoreMessageManager should position a property for non inline attachment detection
 - https://github.com/chibenwa/james-project/commit/fb8b62a387f70c43204214e62d64ad506ecd84ad Rely on property positioned by MessageManager for indexiing HasAttachment
 - https://github.com/chibenwa/james-project/commit/6fc1b2b0c5edb285ffaf2e9672ca1dc14bae575f PropertyBuilder some more tests

The rationals behind this proposal : **not all the mailbox implementations store Attachments** (eg : JPA), however it would be cool that all the search back-end handle this new criterion, **whatever the mailbox storage backend**. As all mailbox backends retireves attachments before storing them, we can compute if we have attachment at that time. Hopefully, the property "hasAttachment" of a message is Immutable. The way I thought for doing this is to use "Properties" which seem to be the least invasive way. Other benefit : we guaranty **cross "search backend" consistency**.

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 1 above)
Comment (mbaechler) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: @chibenwa : you can very well serialize a method result of a bean into an ES document field, I was not arguing to not store hasAttachment at all.

@strokyl : ok

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 4 above)
Comment (mbaechler) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: they could be readable but the real argument IMO is that a ternary operator is better in its own function.

Comment (strokyl) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: I will change to if else, but honestly on this case I am not convinced that it will be more readable than my current ternary operator.

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java [hunk 8]
```diff
@@ -594,10 +595,18 @@ public static Criterion flagSet(Flag flag, boolean isSet) {
         return result;
     }
 
+    public static Criterion hasAttachment() {
+        return new AttachmentCriterion(BooleanOperator.set());
+    }
+
+    public static Criterion hasNoAttachment() {
```
Comment (strokyl) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java]: The work of handling both is almost free in all implementation. I proposed two method hasAttachment and hasNoAttachment mimicing hasFlag and hasNoFlag, to have a consistent API. As a potential futur user of this API, I think it's nice to have a method directly for the negation

File: mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java (hunk 6 above)
Comment (strokyl) [inline mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java]: I will sorry, I kind of rush yesterday pushing because it was important for me to have a first review for this day

File: mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java [hunk 9]
```diff
@@ -971,45 +989,49 @@ private Query createModSeqQuery(SearchQuery.ModSeqCriterion crit) throws Unsuppo
             throw new UnsupportedSearchException();
         }
     }
-    
+
+    private Query createAttachmentQuery(boolean isSet, Query inMailboxes) throws MailboxException, UnsupportedSearchException {
+        return new TermQuery(new Term(HAS_ATTACHMENT_FIELD,Boolean.toString(isSet)));
```
Comment (quynhn) [inline mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java]: one more space after comma ","

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 4 above)
Comment (chibenwa) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: You know, team habits ...

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java (hunk 8 above)
Comment (chibenwa) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java]: Ok, nice ;-)

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 4 above)
Comment (rouazana) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: s/habits/code style/ :)

**Comment by chibenwa** [comment]
2 test failure in MailboxMessageToElasticSearchJsonTest

Feel free to modify the concerned JSON if need be, it's related to the expacations of the ticket

**Comment by chibenwa** [comment]
One more fix related to my changesets https://github.com/chibenwa/james-project/commit/84cf73d6768c07a9e32c94a74ecbb1c6d64bb86d

**Comment by chibenwa** [comment]
I guess you can find several interresting commits in there (last ones are not mentioned here for cherry-pick)

https://github.com/linagora/james-project/pull/615

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 4 above)
Comment (mbaechler) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: @strokyl : nobody argue that you should do an if/else in place. I argue that such an expression deserves a named method. Do you want to say that naming method is not more readable ?

**Comment by chibenwa** [comment]
```
[Building Java 6 james-420_hasattachment-ef6bbe61-dcb5-448f-9cf9-78e5582b3f49] unable to prepare context: unable to evaluate symlinks in context path: lstat /var/jenkins_home/workspace/workflow@4/dockerfiles: no such file or directory
```

I don't remember what that is...

**Comment by chibenwa** [comment]
Test this please

**Comment by chibenwa** [comment]
test this please

**Comment by chibenwa** [comment]
```
[Building Java 8 james-420_hasattachment-5bea1f2c-61a6-426b-9f89-312976fa6291] [ERROR] /james-project/mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java:533: Undefined reference: java.util.stream.Stream java.util.List.stream()
[Building Java 8 james-420_hasattachment-5bea1f2c-61a6-426b-9f89-312976fa6291] [ERROR] /james-project/mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java:533: Undefined reference: boolean java.util.stream.Stream.anyMatch(java.util.function.Predicate)
```

Normal that CI is unhappy with this :-p

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 2 above)
Comment (strokyl) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: Done

File: mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java (hunk 3 above)
Comment (strokyl) [inline mailbox/store/src/main/java/org/apache/james/mailbox/store/search/MessageSearches.java]: This lline is not here anymore so DONE

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java (hunk 4 above)
Comment (strokyl) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: done

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java (hunk 5 above)
Comment (strokyl) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java]: Done

File: mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java [hunk 10]
```diff
@@ -1308,14 +1327,59 @@ public String toString() {
...
     }
 
+    /***
+     * Filter on attachment presence
+     */
+    public static class AttachmentCriterion extends Criterion {
+        private final BooleanOperator operator;
+
+        private AttachmentCriterion(BooleanOperator operator) {
+            super();
```
Comment (strokyl) [inline mailbox/api/src/main/java/org/apache/james/mailbox/model/SearchQuery.java]: Done

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java [hunk 11]
```diff
@@ -68,12 +68,12 @@ public static IndexableMessage from(MailboxMessage message, List<User> users, Te
     }
 
     private void setFlattenedAttachments(MimePart parsingResult, IndexAttachments indexAttachments) {
-        if (indexAttachments.equals(IndexAttachments.YES)) {
-            attachments = parsingResult.getAttachmentsStream()
-                    .collect(Collectors.toList());
-        } else {
-            attachments = ImmutableList.of();
-        }
+        List<MimePart> mimeparts = parsingResult.getAttachmentsStream()
```
Comment (strokyl) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: done

File: mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java [hunk 12]
```diff
@@ -114,6 +114,12 @@ private static ZonedDateTime getSanitizedInternalDate(MailboxMessage message, Zo
             zoneId);
     }
 
+    private static boolean isAttachmentMimepart(MimePart mimepart) {
+        return mimepart.getContentDisposition()
+                .map("attachment"::equals)
```
Comment (strokyl) [inline mailbox/elasticsearch/src/main/java/org/apache/james/mailbox/elasticsearch/json/IndexableMessage.java]: Not here anymore so DONE

File: mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java (hunk 6 above)
Comment (strokyl) [inline mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java]: I remove all space change from my PR so DONE

File: mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java (hunk 7 above)
Comment (strokyl) [inline mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java]: Not here anymore so done

Comment (strokyl) [inline mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java]: Not here anymore so done

File: mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java (hunk 9 above)
Comment (strokyl) [inline mailbox/lucene/src/main/java/org/apache/james/mailbox/lucene/search/LuceneMessageSearchIndex.java]: done

**Comment by aduprat** [comment]
Issue number is wrong, i guess.

**Comment by strokyl** [comment]
test this please

**Comment by strokyl** [comment]
@aduprat This is the issue : https://aduprat.framaboard.org/?controller=TaskViewController&action=show&task_id=420&project_id=6 so 420 isn't correct ?

**Comment by rouazana** [comment]
Nope, we almost never reference Framabord issue numbers, but only Apache's JIRA ones.

**Comment by strokyl** [comment]
Because I put a bad ticket numver, I have redone this PR with the correct number here https://github.com/linagora/james-project/pull/619
